docs/tutorial/programming.md

```diff
@@ -1,10 +1,295 @@
+```
+The `MODE` and `LOOP_ITER` are found in the `data.h`. The `LOOP_ITER` is the effective number of loops while considering the spatial parallelism. For example, if we have 100 elements (each is 64-bit wide), then with a spatial parallelism of 4 PEs, we only need to process 25 cycles so the `LOOP_ITER` is 25. 
+
+The last register written on address `0x3ce` is the start signal of the accelerator. Once started, the accelerator and streamer work together to process the data. We can monitor and poll the busy status of the accelerator by reading on register address `0x3cf`. The busy signal is high when the accelerator is still processing data.
+
+```C
+// Do this to poll the accelerator
+while (read_csr(0x3cf)) {
+};
+```
+
+Once complete, we can check and compare the results in the TCDM and the golden answer stored in L2 memory. The golden output `OUT` is declared in `data.h`.
+
+```C
+// Compare results and check if the
+// accelerator returns correct answers
+// For every incorrect answer, increment err
+uint64_t check_val;
+
+for (uint32_t i = 0; i < DATA_LEN; i++) {
+    // Need to combine upper 64bit bank
+    // with the lower 64 bit bank
+    check_val = *(local_o + i * 2) + *(local_o + i * 2 + 1);
+    if (check_val != OUT[i]) {
+        err++;
+    }
+}
+```
+
+Finally, we can print some performance counters at the end:
+
+```C
+// Read performance counter
+uint32_t perf_count = read_csr(0x3d0);
+
+printf("Accelerator Done! \n");
+printf("Accelerator Cycles: %d \n", perf_count);
+printf("Number of errors: %d \n", err);
+
+```
+
+In summary, you only need to first preload the data into the local TCDM, configure the CSR registers, wait for the accelerator to finish, then finally profile the performance!
+
+# Generating Data
+
+The data processed is defined in the `data.h` included file. Inside the `./target/snitch_cluster/sw/apps/snax-alu/data/data.h`. This file is only generated when you run the `./target/snitch_cluster/sw/apps/snax-alu/data/datagen.py` Python script. To generate data you can run:
+
+```bash
+./datagen.py --mode=0 --length=20 > data.h
+```
+
+The `datagen.py` runs and prints the data into the terminal IO and therefore we need to dump it into the `data.h` file. The `--mode` pertains to what kind of kernel to use in the accelerator. 0 is for add, 1 is for sub, 2 is for mul, and 3 is for XOR. The `--length` pertains to the data lengths.
+
+After generating the script you can investigate the contents of the `data.h`. You should see:
+
+- `MODE` - is the mode that the accelerator will use.
+- `DATA_LEN` - is the length of data to be processed.
+- `LOOP_ITER` - is the data length divided by the amount of parallelism.
+- `A` and `B` - are the input data.
+- `OUT` - is the golden output to which the outputs are compared to.
+
+!!! note
+
+    Generating the `data.h` is already automatic by the `Makefile` we have. Therefore you don't need to run the `datagen.py` manually before building the program.
+
+# Modifying Makefiles
```

Comment:
```suggestion
## Modifying Makefiles
```

Reply:
I think this is a separate main section 😄